src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java

```diff
@@ -0,0 +1,144 @@
+package com.atlassian.bitbucket.jenkins.internal.model.deployment;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.commons.lang3.StringUtils.stripToNull;
+
+/**
+ * The details of an environment that was deployed to.
+ *
+ * @since deployments
+ */
+public final class BitbucketDeploymentEnvironment {
+
+    private static final String DISPLAY_NAME = "displayName";
+    private static final String KEY = "key";
+    private static final String TYPE = "type";
+    private static final String URL = "url";
+
+    private final String key;
+    private final String name;
+    private final String type;
+    private final String url;
+
+    private BitbucketDeploymentEnvironment(Builder builder) {
```

Comment:
The builder offers very little benefit for us here. It is 4 fields. Thus I think we should remove the builder completely and just telescope the constructor. We typically do not use builder in entity classes (I know we have some). 
If you very strongly want to keep it, remove this constructor and call the telescoped one below from the `build`method so we only have one constructor.

Reply:
>Is there a reason you would avoid it other than 'we typically do not use' them?

The builder is more code to maintain, sure not much but it is more. You now have two ways to construct this object.  It is unusual to have builders on entity classes. Depending on how it is done it can cause problems when serialised. If it brings value, it is worth it, but I don't see the value here. The `BitbucketDeployment` has more fields, no builder. I know that not all fields are required here, but they end up as `null` anyway so.. just pass `null` in and avoid creating a whole new object to create this one.

>since there are two fields that are optional

How would that look? I would expect that those fields are held in variables? So you could just pass them in, and if they're not set they're `null` and things just work?
if the caller has to `if (type != null) { builder.type(type)` then we've only gained more code and no benefit. 
Builders are useful, I just don't think it adds just any benefit in this case.

>You're right that I shouldn't have two constructors setting fields here. Is there a reason you'd prefer telescoping it from the builder instead of me calling this from this constructor?

It depends on the case with just 4 fields, I feel this private constructor "obscures" the one that gets called. Having a single constructor is preferable. If we need to have more than one constructor chaining them is preferred. 

Like I've said many times, I just fail to see the value this builder brings. Jenkins is **very** `null` heavy, so passing `null` for unset fields is what is done, in this plugin and other places. I feel this is just more code to maintain. 

Despite the near novel here, if you feel it is valuable I'm happy enough to leave it in.